util.py

```diff
@@ -68,6 +68,85 @@ def _order_data_to_decimal(orders: List[Dict]) -> List[Dict]:
     return orders
 
 
+def restrict_order_sell_amounts_by_balances(
+        orders: List[Dict],
+        accounts: Dict[str, Dict[str, int]]) -> List[Dict]:
+    """Restrict order sell amounts to available account balances.
+
+    This method also filters out orders that end up with a sell amount of zero.
+
+    Args:
+        orders: List of orders.
+        accounts: Dict of accounts and their token balances.
+
+    Returns:
+        The capped orders.
+
+    """
+    orders_capped = []
+
+    # Init dict for remaining balance per account and token pair.
+    remaining_balances = {}
+
+    # Iterate over orders sorted by their limit price [best-to-worst]
+    # (adding +1 in case some number is equal to zero, preserving the ordering).
+    # Potential side-effect:
+    # This may in certain cases interfere with the max-nr-exec-orders or the
+    # min-avg-fee-per-order (economic viability) constraint, where a larger order
+    # with a worse price might be preferred over a smaller order with a better price.
+    for o in sorted(
+            orders,
+            key=lambda o: (Decimal(o['sellAmount']) + 1) / (Decimal(o['buyAmount']) + 1),
+            reverse=True
+    ):
+        tS, tB = o['sellToken'], o['buyToken']
+
+        # Get sell amount (capped by available account balance).
+        sell_amount_old = Decimal(o['sellAmount'])
+
+        # Init remaining balance for new token pair on some account.
+        aID = o['accountID']
+        oID = "%s|%s" % (aID, o.get('orderID'))
+        if (aID, tS, tB) not in remaining_balances:
+            sell_token_balance = Decimal(accounts.get(aID, {}).get(tS, 0))
+            remaining_balances[(aID, tS, tB)] = sell_token_balance
+
+        sell_amount_new = min(sell_amount_old, remaining_balances[aID, tS, tB])
+
+        # Update remaining balance.
+        remaining_balances[aID, tS, tB] -= sell_amount_new
+        assert remaining_balances[aID, tS, tB] >= 0
+
+        logging.debug(
+            "Capping sell amount of <%s> by account balance [%s] : %40d --> %25d"
+            % (oID, tS, sell_amount_old, sell_amount_new))
+
+        # Skip orders with zero sell amount.
+        if sell_amount_new == 0:
+            logging.debug(
+                "Removing order <%s> : zero sell amount or available balance!" % oID)
+            continue
```

Comment:
If we just continue here, are the orders really taken out of the orderbook?

Reply:
Yes - they are not appended to the `orders` list that is the return value.